### PR TITLE
init Storage2UriMapper: ensure JavaCore is initialized after load

### LINKED ISSUE #1835: Storage2UriMapperJavaImpl::doInitializeCache causes needless rebuilds after restarting Eclipse (2021-12)
So I do have annoying runs of the auto build here for ~10min when restarting Eclipse despite everything was built up to date just before the restart - on Linux with our application having >500 projects and >50k source files:
1. Have each (auto) builder up to date, as in: repeated disable+enable auto build does not build anything anymore.
2. Leave auto build enabled.
3. Restart Eclipse 2021-12 (not discovered with 2021-06).
4. See the auto build running for annoying long duration.

The culprit is that "Storage2UriMapperJavaImpl::doInitializeCache" thread may run before the "Initializing Java Tooling" thread.
Because of being early, project information may not be restored from saved Eclipse state yet.
Hence, projects are set up from scratch to some degree, ending up with calls to `Project.touch()`.
The result is a modified `ResourceInfo.charsetAndContentId` on the working tree projects, causing builders to be out of date for no good reason.

When using a breakpoint in `Storage2UriMapperJavaImpl.doInitializeCache` to wait for the "Initializing Java Tooling" thread to finish, `Project.touch()` is not called any more, and the builders remain up to date as restored from saved Eclipse state.

Stack trace of `ResourceInfo.incrementContentId()` being called in "Storage2UriMapperJavaImpl::doInitializeCache" thread:
> "Storage2UriMapperJavaImpl::doInitializeCache" #48 prio=6 os_prio=0 cpu=2534.78ms elapsed=48.41s tid=0x00007fc58f5dd000 nid=0x3bbb5 at breakpoint [0x00007fc4d8281000]
   java.lang.Thread.State: RUNNABLE
        at org.eclipse.core.internal.resources.ResourceInfo.incrementContentId(ResourceInfo.java:287)
        at org.eclipse.core.internal.resources.Resource.touch(Resource.java:1759)
        at org.eclipse.core.internal.resources.Project.touch(Project.java:1323)
        at org.eclipse.jdt.internal.core.SetContainerOperation.executeOperation(SetContainerOperation.java:122)
        at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:740)
        at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2313)
        at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2338)
        at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:811)
        at org.eclipse.jdt.internal.core.JavaModelManager.getClasspathContainer(JavaModelManager.java:2095)
        at org.eclipse.jdt.core.JavaCore.getClasspathContainer(JavaCore.java:3783)
        at org.eclipse.jdt.internal.core.JavaProject.resolveClasspath(JavaProject.java:3240)
        at org.eclipse.jdt.internal.core.JavaProject.resolveClasspath(JavaProject.java:3404)
        at org.eclipse.jdt.internal.core.JavaProject.getResolvedClasspath(JavaProject.java:2489)
        at org.eclipse.jdt.internal.core.JavaProject.buildStructure(JavaProject.java:535)
        at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:266)
        at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:597)
        at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:328)
        at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:314)
        at org.eclipse.jdt.internal.core.JavaElement.getChildren(JavaElement.java:269)
        at org.eclipse.jdt.internal.core.JavaProject.getPackageFragmentRoots(JavaProject.java:2353)
        at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.updateCache(Storage2UriMapperJavaImpl.java:478)
        at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.access$7(Storage2UriMapperJavaImpl.java:474)
        at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl$4.run(Storage2UriMapperJavaImpl.java:674)
        at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2313)
        at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2338)
        at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.doInitializeCache(Storage2UriMapperJavaImpl.java:686)
        at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.access$3(Storage2UriMapperJavaImpl.java:665)
        at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl$3.run(Storage2UriMapperJavaImpl.java:654)
        at java.lang.Thread.run(java.base@11.0.14/Thread.java:829)

### PR DESCRIPTION
**Pull request by haubi**
Fixes #1835

Signed-off-by: Michael Haubenwallner <michael.haubenwallner@ssi-schaefer.com>

### CODE REVIEW

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperJavaImpl.java [hunk 1]
```diff
@@ -669,6 +669,7 @@ protected void doInitializeCache() throws CoreException {
 				@Override
 				public void run(IProgressMonitor monitor) throws CoreException {
 					if(!isInitialized) {
+						JavaCore.initializeAfterLoad(monitor);
```
Comment (cdietrich) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperJavaImpl.java:672]: can this lead to deadlocks?

**Comment by cdietrich** [comment]
@iloveeclipse what do you think?

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperJavaImpl.java (hunk 1 above)
Comment (iloveeclipse) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperJavaImpl.java:672]: I would avoid calling internal init of JDT here, that is really not supposed to be done by 3rd party code.
Sounds like a hack.

**Comment by cdietrich** [comment]
is this obsolete now ?!?

**Comment by haubi** [comment]
Yes, does seem so, thanks!

**Pull request closed without merging**